pkg/internal/cnsoperator/cnsfilevolumeclient/cnsfilevolumeclient.go

```diff
@@ -0,0 +1,251 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cnsfilevolumeclient
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	"k8s.io/apimachinery/pkg/api/errors"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/tools/cache"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"sigs.k8s.io/vsphere-csi-driver/pkg/csi/service/logger"
+	"sigs.k8s.io/vsphere-csi-driver/pkg/internal/cnsoperator/cnsfilevolumeclient/v1alpha1"
+	k8s "sigs.k8s.io/vsphere-csi-driver/pkg/kubernetes"
+)
+
+// FileVolumeClient exposes an interface to support
+// configuration of CNS file volume ACL's.
+type FileVolumeClient interface {
+	// AddClientVMToIPList adds the input clientVMName to the list of
+	// clientVMNames that expose the same external clientVMIP for a
+	// given file volume. fileVolumeName is used to uniquely
+	// identify CnsFileVolumeClient instances.
+	AddClientVMToIPList(ctx context.Context, fileVolumeName, clientVMName, clientVMIP string) error
+	// RemoveClientVMFromIPList removes the input clientVMName from
+	// the list of clientVMNames that expose the same external
+	// clientVMIP for a given file volume. fileVolumeName is used
+	// to uniquely identify CnsFileVolumeClient instances.
+	RemoveClientVMFromIPList(ctx context.Context, fileVolumeName, clientVMName, clientVMIP string) error
+}
+
+// fileVolumeClient maintains a client to the API
+// server for operations on CnsFileVolumeClient instance.
+// It also contains a per instance lock to handle
+// concurrent operations.
+type fileVolumeClient struct {
+	client client.Client
+	// Per volume lock for concurrent access to CnsFileVolumeClient instances.
+	// Keys are strings representing volume handles (or SV-PVC names).
+	// Values are individual sync.Mutex locks that need to be held
+	// to make updates to the CnsFileVolumeClient instance on the API server.
+	volumeLock *sync.Map
+}
+
+var (
+	fileVolumeClientInstanceLock sync.Mutex
+	fileVolumeClientInstance     *fileVolumeClient
+)
+
+// GetFileVolumeClientInstance returns a singleton of type FileVolumeClient.
+// Initializes the singleton if not already initialized.
+func GetFileVolumeClientInstance(ctx context.Context) (FileVolumeClient, error) {
+	fileVolumeClientInstanceLock.Lock()
+	defer fileVolumeClientInstanceLock.Unlock()
+	if fileVolumeClientInstance == nil {
+		log := logger.GetLogger(ctx)
+		config, err := k8s.GetKubeConfig(ctx)
+		if err != nil {
+			log.Errorf("failed to get kubeconfig. Err: %v", err)
+			return nil, err
+		}
+		k8sclient, err := k8s.NewClientForGroup(ctx, config, v1alpha1.GroupVersion.Group)
```

Comment:
How is the k8s client recovered if the creds are updated? Dont we need a way to reload the client when the creds are updated?

Reply:
Do we anticipate creds to be updated for the in cluster config? CNS Operator will use these functions to Read/Write CnsFileVolumeClients from the SV API server. I think we've always assumed that this config does not get updated and k8s clients dont need to be reconfigured, so we'd potentially need to make this change in multiple places if the in cluster config is being updated too..

Reply:
Actually I'm not very sure on this. Maybe we should check with the guest cluster manager team. Regardless, fixing this should be a separate change and needs to be done in other places you pointed to. So, its fine for now.